### PR TITLE
FED-3114: Lazy

### PR DESCRIPTION
**Pull request by kealjones-wk**
# Depends on https://github.com/Workiva/react-dart/pull/406

## Motivation
  
We need React.lazy in order to support some new asynchronous component patterns.
 
## Changes
  
- Add React.lazy
- Add a lot of test coverage

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
      - [ ] Run the suspense example and ensure it is working correctly still
      - [ ] CI Passes
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: test/over_react/component/suspense_component_test.dart [hunk 1]
```diff
@@ -15,46 +15,14 @@
...
-          ),
-        ),
-      ),
-    ),
-    factoryConfig,
-  );
-}
-
 const lazyId = 'lazy';
 const gregIsNotLazy = 'gregisnotlazy';
```
Comment (kealjones-wk) [inline test/over_react/component/suspense_component_test.dart:27]: lol

File: lib/src/util/lazy.dart [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  // ignore: invalid_use_of_protected_member
+  var propsFactory = _config.propsFactory;
+
+  final lazyFactoryProxy = react.lazy(() async {
+    final factory = await load();
+    // By using a wrapper uiForwardRef it ensures that we have a matching factory proxy type given to react-dart's lazy,
+    // a `ReactDartWrappedComponentFactoryProxy`. This is necessary to have consistent prop conversions since we don't
+    // have access to the original factory proxy outside of this async block.
+    final wrapper = uiForwardRef<TProps>(
```
Comment (greglittlefield-wf) [inline lib/src/util/lazy.dart:84]: See note [here](https://github.com/Workiva/react-dart/pull/406#discussion_r1779289577) about moving the wrapper into react-dart

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Mainly some comments on tests, otherwise this is looking good!

File: test/over_react/component/lazy_test.dart [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+        final ref = createRef<Element>();
+        final renderErrors = [];
+        bool mapPropFunctionCalled = false;
+        bool functionPropCalled = false;
+
+        render(
+          (components.ErrorBoundary()
+              ..onComponentDidCatch = ((error, _) => renderErrors.add(error))
+              ..shouldLogErrors = false
+              ..fallbackUIRenderer = ((_, __) => Dom.span()('An error occurred during render')))(
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply test/over_react/component/lazy_test.dart:39]: sure: 
Screenshot 2024-09-30 at 3 41 36 PM

[the remark this answers is not in the document]
Comment (kealjones-wk) [reply test/over_react/component/lazy_test.dart:39]: just realized i should probably try on the VPN...

File: lib/src/util/lazy.dart (hunk 2 above)
Comment (kealjones-wk) [reply lib/src/util/lazy.dart:84]: So, i tried removing the wrapper at this level and it caused test failures and wasn't converting props like function props in some of the type assertion tests i added... seems like we need it at both levels. :/

Comment (greglittlefield-wf) [reply lib/src/util/lazy.dart:84]: Hmm interesting. To me it doesn't seem like we'd need it at both levels, unless it's some JS interop tradeoff case that we might not need to support.

Once I get a chance I can try removing it locally and dig into what the specific failures are

Comment (greglittlefield-wf) [reply lib/src/util/lazy.dart:84]: Hm I removed the wrapper and am not seeing any test failures, locally or in CI; here's a PR: https://github.com/Workiva/over_react/pull/953

Oh, is it possible that you were trying with an outdated version of react-dart, from before the wrapper was removed? You can force it to get the latest with `dart pub upgrade react`

Comment (kealjones-wk) [reply lib/src/util/lazy.dart:84]: >Oh, is it possible that you were trying with an outdated version of react-dart, from before the wrapper was removed? You can force it to get the latest with dart pub upgrade react

Crap yes that probably was it... damn.....

File: test/over_react/component/lazy_test.dart [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+        final props = factory(existingMap);
+
+        expect(props.stringProp, equals('test'));
+
+        props.stringProp = 'modified';
+        expect(props.stringProp, equals('modified'));
+        expect(existingMap['stringProp'], equals('modified'));
+      });
+    });
+
```
Comment (greglittlefield-wf) [inline test/over_react/component/lazy_test.dart:102]: It looks like there isn't a basic test in here that the lazy component actually renders, though it is tested in the wrapper cases, and also in suspense_test.dart. 

Would you mind either adding a comment in here that mentions that, or adding a test here?

File: lib/src/util/lazy.dart [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    throw ArgumentError('_config is required when using a custom props class and should be a UiFactoryConfig<TProps>. Make sure you are '
+        r'using either the generated factory config (i.e. _$FooConfig) or manually declaring your config correctly.');
+  }
+  // ignore: invalid_use_of_protected_member
+  var propsFactory = _config.propsFactory;
+
+  final lazyFactoryProxy = react.lazy(() async {
+    final factory = await load();
+    return factory().componentFactory!;
+  });
```
Comment (greglittlefield-wf) [inline lib/src/util/lazy.dart:82]: Aw one thing I just realized is that there's no more `displayName` logic, so the react-dart wrapper just shows up as `Anonymous`

Screenshot 2024-10-04 at 3 05 54 PM

I'd be down to not worry about that in this PR, though, and do that as a follow-up improvement. Your call

File: test/over_react/component/lazy_test.dart [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+    group('does not throw an error when', () {
+      test('config is null, IF it only expects GenericUiProps', () {
+        testLazy() => lazy(
+            () async => uiFunction<UiProps>(
+                  (props) => (Dom.div()..addTestId('testId3'))('id: ${props.id}'),
+                  UiFactoryConfig(),
+                ),
+            null);
+        expect(testLazy, returnsNormally);
+        expect(testLazy(), isA<UiFactory<UiProps>>());
```
Comment (greglittlefield-wf) [inline test/over_react/component/lazy_test.dart:53]: Can we also test the same `initializes the factory variable with a function...` cases as we did below, to make sure this is working properly?

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
A couple more things, should be my final pass!

File: test/over_react/component/lazy_test.dart [hunk 7]
```diff
@@ -0,0 +1,663 @@
...
+
+part 'lazy_test.over_react.g.dart';
+
+main() {
+  enableTestMode();
+
+  group('lazy', () {
+    // The fallback and the lazy component render the same thing...
+    uiPropsTest(
+      uiFunction<UiProps>(
```
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply test/over_react/component/lazy_test.dart:31]: yeah honestly wasn't super happy with this either...

File: lib/src/util/lazy.dart (hunk 5 above)
Comment (kealjones-wk) [reply lib/src/util/lazy.dart:82]: I mean, i could just add like `}, displayName: 'lazy(anonymous)')` to the react-dart side for the time being lol

Comment (kealjones-wk) [reply lib/src/util/lazy.dart:82]: or actually would `'lazy(${factory.type.displayName})'` work in the react-dart side?

Comment (kealjones-wk) [reply lib/src/util/lazy.dart:82]: Screenshot 2024-10-07 at 9 29 40 AM

File: test/over_react/component/lazy_test.dart (hunk 4 above)
Comment (kealjones-wk) [reply test/over_react/component/lazy_test.dart:146]: Added one under the `does not throw an error when` group called `wrapped in suspense (renders correctly)` that checks that there were no errorboundary errors and that the component is displaying correctly. Its also kinda asserted in `renders a component from end to end, successfully reading props via typed getters` from the `uiPropsTests`

File: test/over_react/component/lazy_test.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (kealjones-wk) [reply test/over_react/component/lazy_test.dart]: copy/pasted those tests entirely into this file so that we could test without the wonkiness

**Review by greglittlefield-wf — COMMENTED** [review]
LGTM, just need to get the react-dart changes released

**Review by greglittlefield-wf — APPROVED** [review]
+10 pending passing CI

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**